CONTRIBUTING.md

```diff
@@ -9,70 +36,163 @@ license the code under the same terms as the project's overall 'outbound'
+Synapse. [docs/dev](docs/dev) exists primarily to house documentation for
+Synapse developers. [docs/admin_api](docs/admin_api) houses documentation
+regarding Synapse's Admin API, which is used mostly by sysadmins and external
+service developers.
+
+If you add new files added to either of these folders, please use [Github-Flavoured
+Markdown](https://guides.github.com/features/mastering-markdown/).
+
+Some documentation also exists in [Synapse's Github
+Wiki](https://github.com/matrix-org/synapse/wiki), although this is primarily
+contributed to by community authors.
 
-```
-# Install the dependencies
-pip install -e ".[lint,mypy]"
 
-# Run the linter script
+# 7. Test, test, test!
+<a name="test-test-test"></a>
+
+While you're developing and before submitting a patch, you'll
+want to test your code.
+
+## Run the linters.
+
+The linters look at your code and do two things:
+
+- ensure that your code follows the coding style adopted by the project;
+- catch a number of errors in your code.
+
+They're very fast, don't hesitate!
+
+```sh
+source ./env/bin/activate
 ./scripts-dev/lint.sh
 ```
 
-**Note that the script does not just test/check, but also reformats code, so you
-may wish to ensure any new code is committed first**.
+Note that the linters *will modify your files* to fix styling errors.
+Make sure that you have saved all your files.
 
-By default, this script checks all files and can take some time; if you alter
-only certain files, you might wish to specify paths as arguments to reduce the
-run-time:
 
-```
+If you wish to restrict the linters to only run on some files, you can instead run:
+
+```sh
+source ./env/bin/activate
 ./scripts-dev/lint.sh path/to/file1.py path/to/file2.py path/to/folder
 ```
 
-You can also provide the `-d` option, which will lint the files that have been
-changed since the last git commit. This will often be significantly faster than
-linting the whole codebase.
 
-Before pushing new changes, ensure they don't produce linting errors. Commit any
-files that were corrected.
+## Run the unit tests.
+
+The unit tests run parts of Synapse, including your changes, to see if anything
+was broken. They are slower than the linters but will typically catch more errors.
+
+```sh
+source ./env/bin/activate
+python -m twisted.trial tests
+```
+
+If you wish to only run *some* unit tests, you may specify
+another module instead of `tests`:
+
+```sh
+source ./env/bin/activate
+python -m twisted.trial tests.rest.admin.test_room
+```
+
+If your tests fail, you may wish to look at the logs:
+
+```sh
+less _trial_temp/test.log
+```
+
+## Run the integration tests.
+
+The integration tests run the full Synapse, including your changes, to
+see if anything was broken. They are slower than the unit tests but will
+typically catch more errors.
+
+
+The following will run the test suite using postgres. You will need [Docker](https://docs.docker.com/get-docker/) installed.
+
+```sh
+./test_postgresql.sh
```

Comment:
I don't think this is the integration tests? This looks to me like it is running the unit tests under PostgreSQL? I've never seen this file before though, good to know this is possible!

Reply:
I don't know, I've found this in the documentation, I haven't tested it myself!